File: crates/a1-ai/2.8.0/sdk/typescript/tests/passport.test.ts
import { describe, it, expect, jest, beforeEach } from "@jest/globals";
import {
  PassportClient,
  PassportError,
  PassportReceipt,
  withA1Passport,
  PassportGuard,
} from "../src/passport.js";

// ── Shared fixtures ───────────────────────────────────────────────────────────

const SAMPLE_RECEIPT: PassportReceipt = {
  passport_namespace: "test-agent",
  fingerprint_hex: "de".repeat(32),
  capability_mask_hex: "ff".repeat(32),
  narrowing_commitment_hex: "ab".repeat(32),
  chain_depth: 1,
};

const WRAPPED_RECEIPT = { receipt: SAMPLE_RECEIPT };

function mockFetch(response: object, status = 200): typeof globalThis.fetch {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (jest.fn() as any).mockResolvedValue({
    ok: status < 400,
    status,
    json: async () => response,
  }) as unknown as typeof globalThis.fetch;
}

// ── PassportClient ────────────────────────────────────────────────────────────

describe("PassportClient.authorize", () => {
  let client: PassportClient;

  beforeEach(() => {
    client = new PassportClient("http://localhost:8080");
  });

  it("returns a PassportReceipt on 200 with nested receipt", async () => {
    globalThis.fetch = mockFetch(WRAPPED_RECEIPT);

    const receipt = await client.authorize({
      chain: { certs: [] },
      intent_name: "trade.equity",
      executor_pk_hex: "aa".repeat(32),
    });

    expect(receipt.passport_namespace).toBe("test-agent");
    expect(receipt.chain_depth).toBe(1);
    expect(receipt.fingerprint_hex).toBe("de".repeat(32));
  });

  it("returns a PassportReceipt on 200 with flat receipt", async () => {
    globalThis.fetch = mockFetch(SAMPLE_RECEIPT);

    const receipt = await client.authorize({
      chain: {},
      intent_name: "read.data",
      executor_pk_hex: "",
    });

    expect(receipt.passport_namespace).toBe("test-agent");
    expect(receipt.chain_depth).toBe(1);
  });

  it("throws PassportError on non-200 with structured error body", async () => {
    globalThis.fetch = mockFetch(
      { error: "scope violation", error_code: "SCOPE_VIOLATION" },
      403
    );

    await expect(
      client.authorize({ chain: {}, intent_name: "trade.equity", executor_pk_hex: "" })
    ).rejects.toMatchObject({
      errorCode: "SCOPE_VIOLATION",
      httpStatus: 403,
      message: "scope violation",
    });
  });

  it("throws PassportError on non-200 with no JSON body", async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    globalThis.fetch = (jest.fn() as any).mockResolvedValue({
      ok: false,
      status: 503,
      json: async () => { throw new Error("not json"); },
    }) as unknown as typeof globalThis.fetch;

    await expect(
      client.authorize({ chain: {}, intent_name: "trade.equity", executor_pk_hex: "" })
    ).rejects.toBeInstanceOf(PassportError);
  });

  it("includes intent_params in the request body when provided", async () => {
    let captured: unknown;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    globalThis.fetch = (jest.fn() as any).mockImplementation(async (_url: unknown, init: unknown) => {
      captured = JSON.parse((init as RequestInit).body as string);
      return { ok: true, status: 200, json: async () => WRAPPED_RECEIPT };
    }) as unknown as typeof globalThis.fetch;

    await client.authorize({
      chain: {},
      intent_name: "trade.equity",
      executor_pk_hex: "pk",
      intent_params: { symbol: "AAPL" },
    });

    expect((captured as Record<string, unknown>)["intent_params"]).toEqual({ symbol: "AAPL" });
    expect((captured as Record<string, unknown>)["executor_pk_hex"]).toBe("pk");
  });

  it("respects custom headers and timeoutMs constructor options", () => {
    const custom = new PassportClient("http://host", {
      headers: { Authorization: "Bearer tok" },
      timeoutMs: 5000,
    });
    expect(custom).toBeDefined();
  });
});

// ── PassportError ─────────────────────────────────────────────────────────────

describe("PassportError", () => {
  it("is an instance of Error", () => {
    const e = new PassportError("denied", "DENIED", 403);
    expect(e).toBeInstanceOf(Error);
    expect(e.name).toBe("PassportError");
    expect(e.errorCode).toBe("DENIED");
    expect(e.httpStatus).toBe(403);
    expect(e.message).toBe("denied");
  });

  it("uses defaults when optional args are omitted", () => {
    const e = new PassportError("oops");
    expect(e.errorCode).toBe("PASSPORT_ERROR");
    expect(e.httpStatus).toBe(403);
  });
});

// ── withA1Passport ─────────────────────────────────────────────────────────

describe("withA1Passport", () => {
  type TradeArgs = {
    symbol: string;
    signed_chain: unknown;
    executor_pk_hex?: string;
  };

  it("calls the gateway then executes the wrapped function on success", async () => {
    globalThis.fetch = mockFetch(WRAPPED_RECEIPT);

    const client = new PassportClient("http://localhost:8080");
    const callLog: string[] = [];

    const guarded = withA1Passport(
      async (args: TradeArgs) => {
        callLog.push("executed:" + args.symbol);
        return "filled";
      },
      { client, capability: "trade.equity" }
    );

    const result = await guarded({
      symbol: "AAPL",
      signed_chain: { certs: [] },
      executor_pk_hex: "aa".repeat(32),
    });

    expect(result).toBe("filled");
    expect(callLog).toEqual(["executed:AAPL"]);
  });

  it("blocks the wrapped function when gateway rejects", async () => {
    globalThis.fetch = mockFetch({ error: "denied", error_code: "SCOPE_VIOLATION" }, 403);

    const client = new PassportClient("http://localhost:8080");
    const callLog: string[] = [];

    const guarded = withA1Passport(
      async (_args: TradeArgs) => {
        callLog.push("should-not-run");
        return "bad";
      },
      { client, capability: "trade.equity" }
    );

    await expect(
      guarded({ symbol: "AAPL", signed_chain: {} })
    ).rejects.toBeInstanceOf(PassportError);

    expect(callLog).toHaveLength(0);
  });

  it("throws PassportError when signed_chain is missing", async () => {
    const client = new PassportClient("http://localhost:8080");

    const guarded = withA1Passport(
      async (_args: Record<string, unknown>) => "ok",
      { client, capability: "trade.equity", chainKey: "signed_chain" }
    );

    await expect(guarded({ symbol: "AAPL" })).rejects.toMatchObject({
      errorCode: "MISSING_CHAIN",
    });
  });

  it("respects custom chainKey and executorKey options", async () => {
    let capturedChain: unknown;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    globalThis.fetch = (jest.fn() as any).mockImplementation(async (_url: unknown, init: unknown) => {
      capturedChain = JSON.parse((init as RequestInit).body as string);
      return { ok: true, status: 200, json: async () => WRAPPED_RECEIPT };
    }) as unknown as typeof globalThis.fetch;

    const client = new PassportClient("http://localhost:8080");

    const guarded = withA1Passport(
      async (_args: Record<string, unknown>) => "ok",
      { client, capability: "trade.equity", chainKey: "myChain", executorKey: "myPK" }
    );

    await guarded({ myChain: { certs: [] }, myPK: "pk123" });

    const body = capturedChain as Record<string, unknown>;
    expect(body["executor_pk_hex"]).toBe("pk123");
    expect(body["chain"]).toEqual({ certs: [] });
  });

  it("passes the original arguments unchanged to the wrapped function", async () => {
    globalThis.fetch = mockFetch(WRAPPED_RECEIPT);
    const client = new PassportClient("http://localhost:8080");

    let received: TradeArgs | undefined;

    const guarded = withA1Passport(
      async (args: TradeArgs) => {
        received = args;
        return "ok";
      },
      { client, capability: "trade.equity" }
    );

    const input: TradeArgs = { symbol: "GOOG", signed_chain: { v: 1 }, executor_pk_hex: "ee" };
    await guarded(input);

    expect(received).toEqual(input);
  });
});

// ── PassportGuard decorator ───────────────────────────────────────────────────

describe("PassportGuard class-method decorator", () => {
  it("is exported and callable", () => {
    expect(typeof PassportGuard).toBe("function");
  });
});

File: crates/a1-ai/2.8.0/sdk/go/kya/passport_test.go
package a1

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

// ── Fixtures ──────────────────────────────────────────────────────────────────

var samplePassportReceipt = map[string]any{
	"receipt": map[string]any{
		"passport_namespace":       "test-agent",
		"fingerprint_hex":          "deadbeef" + "deadbeef" + "deadbeef" + "deadbeef" + "deadbeef" + "deadbeef" + "deadbeef" + "deadbeef",
		"capability_mask_hex":      "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff" + "ff",
		"narrowing_commitment_hex": "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab" + "ab",
		"chain_depth":              1,
	},
}

func passportOkServer(t *testing.T) *httptest.Server {
	t.Helper()
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(samplePassportReceipt)
	}))
}

func passportErrServer(t *testing.T, status int, code string) *httptest.Server {
	t.Helper()
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(status)
		_ = json.NewEncoder(w).Encode(map[string]string{
			"error":      "authorization denied",
			"error_code": code,
		})
	}))
}

// ── AuthorizePassport ─────────────────────────────────────────────────────────

func TestAuthorizePassport_Success(t *testing.T) {
	srv := passportOkServer(t)
	defer srv.Close()

	client := New(srv.URL)
	receipt, err := client.AuthorizePassport(
		context.Background(),
		map[string]any{"certs": []any{}},
		"trade.equity",
		"aabbcc",
		nil,
	)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if receipt.PassportNamespace != "test-agent" {
		t.Errorf("namespace: got %q, want %q", receipt.PassportNamespace, "test-agent")
	}
	if receipt.ChainDepth != 1 {
		t.Errorf("chain_depth: got %d, want 1", receipt.ChainDepth)
	}
}

func TestAuthorizePassport_FlatReceiptFallback(t *testing.T) {
	flat := map[string]any{
		"passport_namespace":       "bot",
		"fingerprint_hex":          "00" + "00" + "00" + "00",
		"capability_mask_hex":      "ff" + "ff",
		"narrowing_commitment_hex": "cc" + "cc",
		"chain_depth":              2,
	}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(flat)
	}))
	defer srv.Close()

	client := New(srv.URL)
	receipt, err := client.AuthorizePassport(context.Background(), nil, "read.data", "", nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if receipt.PassportNamespace != "bot" {
		t.Errorf("namespace: got %q, want %q", receipt.PassportNamespace, "bot")
	}
	if receipt.ChainDepth != 2 {
		t.Errorf("chain_depth: got %d, want 2", receipt.ChainDepth)
	}
}

func TestAuthorizePassport_AuthorizationFailure(t *testing.T) {
	srv := passportErrServer(t, http.StatusForbidden, "PASSPORT_NARROWING_VIOLATION")
	defer srv.Close()

	client := New(srv.URL)
	_, err := client.AuthorizePassport(context.Background(), nil, "trade.equity", "", nil)
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	pe, ok := err.(*PassportError)
	if !ok {
		t.Fatalf("expected *PassportError, got %T: %v", err, err)
	}
	if pe.ErrorCode != "PASSPORT_NARROWING_VIOLATION" {
		t.Errorf("error_code: got %q, want %q", pe.ErrorCode, "PASSPORT_NARROWING_VIOLATION")
	}
	if pe.Status != http.StatusForbidden {
		t.Errorf("status: got %d, want %d", pe.Status, http.StatusForbidden)
	}
}

func TestAuthorizePassport_NilIntentParams_Allowed(t *testing.T) {
	srv := passportOkServer(t)
	defer srv.Close()

	client := New(srv.URL)
	_, err := client.AuthorizePassport(context.Background(), nil, "trade.equity", "", nil)
	if err != nil {
		t.Fatalf("nil intent_params should not error: %v", err)
	}
}

func TestAuthorizePassport_WithIntentParams(t *testing.T) {
	var gotBody map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_ = json.NewDecoder(r.Body).Decode(&gotBody)
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(samplePassportReceipt)
	}))
	defer srv.Close()

	client := New(srv.URL)
	params := map[string]any{"symbol": "AAPL", "qty": 10}
	_, err := client.AuthorizePassport(context.Background(), nil, "trade.equity", "pk", params)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if gotBody["intent_name"] != "trade.equity" {
		t.Errorf("intent_name: got %v", gotBody["intent_name"])
	}
	if gotBody["executor_pk_hex"] != "pk" {
		t.Errorf("executor_pk_hex: got %v", gotBody["executor_pk_hex"])
	}
}

// ── WithPassport ──────────────────────────────────────────────────────────────

type tradeArgs struct {
	Symbol        string `json:"symbol"`
	SignedChain   any    `json:"signed_chain"`
	ExecutorPKHex string `json:"executor_pk_hex"`
}

func TestWithPassport_SuccessCallsWrappedFn(t *testing.T) {
	srv := passportOkServer(t)
	defer srv.Close()

	client := New(srv.URL)
	called := false

	guardedFn := WithA1Passport(client, func(ctx context.Context, args tradeArgs) (string, error) {
		called = true
		return "filled", nil
	}, PassportOptions{Capability: "trade.equity"})

	result, err := guardedFn(context.Background(), tradeArgs{
		Symbol:        "AAPL",
		SignedChain:   map[string]any{"certs": []any{}},
		ExecutorPKHex: "aabb",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result != "filled" {
		t.Errorf("result: got %q, want %q", result, "filled")
	}
	if !called {
		t.Error("wrapped function was not called")
	}
}

func TestWithPassport_AuthFailureBlocksWrappedFn(t *testing.T) {
	srv := passportErrServer(t, http.StatusForbidden, "SCOPE_VIOLATION")
	defer srv.Close()

	client := New(srv.URL)
	called := false

	guardedFn := WithA1Passport(client, func(ctx context.Context, args tradeArgs) (string, error) {
		called = true
		return "should-not-run", nil
	}, PassportOptions{Capability: "trade.equity"})

	_, err := guardedFn(context.Background(), tradeArgs{Symbol: "AAPL"})
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if called {
		t.Error("wrapped function must not run when authorization fails")
	}
	pe, ok := err.(*PassportError)
	if !ok {
		t.Fatalf("expected *PassportError, got %T", err)
	}
	if pe.ErrorCode != "SCOPE_VIOLATION" {
		t.Errorf("error_code: got %q", pe.ErrorCode)
	}
}

func TestWithPassport_MissingChainField(t *testing.T) {
	client := New("http://localhost:9999") // no server needed

	type noChainArgs struct {
		Symbol string `json:"symbol"`
	}

	guardedFn := WithA1Passport(client, func(ctx context.Context, args noChainArgs) (string, error) {
		return "ok", nil
	}, PassportOptions{Capability: "trade.equity"})

	_, err := guardedFn(context.Background(), noChainArgs{Symbol: "AAPL"})
	if err == nil {
		t.Fatal("expected error for missing chain field")
	}
	pe, ok := err.(*PassportError)
	if !ok {
		t.Fatalf("expected *PassportError, got %T: %v", err, err)
	}
	if pe.ErrorCode != "MISSING_CHAIN" {
		t.Errorf("error_code: got %q, want MISSING_CHAIN", pe.ErrorCode)
	}
}

func TestWithPassport_CustomFieldNames(t *testing.T) {
	srv := passportOkServer(t)
	defer srv.Close()

	type customArgs struct {
		MyChain any    `json:"chain"`
		MyPK    string `json:"pk"`
	}

	client := New(srv.URL)
	called := false

	guardedFn := WithA1Passport(client, func(ctx context.Context, args customArgs) (string, error) {
		called = true
		return "ok", nil
	}, PassportOptions{
		Capability:    "trade.equity",
		ChainField:    "MyChain",
		ExecutorField: "MyPK",
	})

	_, err := guardedFn(context.Background(), customArgs{
		MyChain: map[string]any{},
		MyPK:    "pk",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !called {
		t.Error("wrapped function was not called")
	}
}

func TestPassportError_ErrorString(t *testing.T) {
	e := &PassportError{Message: "denied", ErrorCode: "SCOPE_VIOLATION", Status: 403}
	s := e.Error()
	if s == "" {
		t.Error("Error() must not return empty string")
	}
}